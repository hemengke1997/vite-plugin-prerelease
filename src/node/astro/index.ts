import { type AstroIntegration } from 'astro'
import { prerelease as main, type Options } from '..'

export function prerelease(options?: Pick<Options, 'excludeEnvs' | 'prereleaseEnv' | 'prereleaseWidget'>): any {
  const { prereleaseEnv = 'production' } = options || {}
  return {
    name: 'vite-plugin-prerelease-integration',
    hooks: {
      'astro:config:setup': ({ injectScript, updateConfig }) => {
        updateConfig({
          vite: {
            plugins: [
              main({
                ...options,
                mode: 'runtime',
                entry: 'astro:scripts/page.js',
              }),
            ],
          },
        })

        if (process.env.NODE_ENV === prereleaseEnv || process.env.NODE_ENV === 'production') {
          return
        }
        injectScript('page', '')
      },
    },
  } as AstroIntegration
}
