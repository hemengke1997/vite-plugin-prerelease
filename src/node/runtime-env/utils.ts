import { parse } from 'dotenv'
import fs from 'node:fs'
import path from 'node:path'
import { normalizePath, type ResolvedConfig } from 'vite'

function arraify<T>(target: T | T[]): T[] {
  return Array.isArray(target) ? target : [target]
}

function getEnvFilesForMode(mode: string, envDir: string): string[] {
  return [
    /** default file */ `.env`,
    /** local file */ `.env.local`,
    /** mode file */ `.env.${mode}`,
    /** mode local file */ `.env.${mode}.local`,
  ].map((file) => normalizePath(path.join(envDir, file)))
}

function tryStatSync(file: string): fs.Stats | undefined {
  try {
    // The "throwIfNoEntry" is a performance optimization for cases where the file does not exist
    return fs.statSync(file, { throwIfNoEntry: false })
  } catch {
    // Ignore errors
  }
}

export function loadEnv(mode: string, envDir: string, prefixes: string | string[] = 'VITE_'): Record<string, string> {
  prefixes = arraify(prefixes)
  const env: Record<string, string> = {}
  const envFiles = getEnvFilesForMode(mode, envDir)

  const parsed = Object.fromEntries(
    envFiles.flatMap((filePath) => {
      if (!tryStatSync(filePath)?.isFile()) return []

      return Object.entries(parse(fs.readFileSync(filePath)))
    }),
  )

  // only keys that start with prefix are exposed to client
  for (const [key, value] of Object.entries(parsed)) {
    if (prefixes.some((prefix) => key.startsWith(prefix))) {
      env[key] = value
    }
  }

  return env
}

export function resolveEnvFromConfig(config: ResolvedConfig, prereleaseEnv: string) {
  const env = {
    prerelease: {
      ...config.env,
      ...loadEnv(prereleaseEnv, config.envDir, config.envPrefix),
      // define import.meta.env.PRERELEASE
      PRERELEASE: true,
      MODE: prereleaseEnv,
    },
    current: {
      ...config.env,
      ...loadEnv(config.mode, config.envDir, config.envPrefix),
    },
  }

  global.__env__ = env

  return env
}
