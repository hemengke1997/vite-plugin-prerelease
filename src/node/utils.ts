import * as cheerio from 'cheerio'
import { type Element } from 'domhandler'

export function transformHtml(originHtml: string, prereleaseHtml: string) {
  const $ = cheerio.load(originHtml)
  const pre$ = cheerio.load(prereleaseHtml)

  const generateDynamicTags = ($: cheerio.CheerioAPI) => {
    return [
      ...$('script').filter((_, el) => Boolean($(el).attr('src'))),
      ...$('link').filter((_, el) => Boolean($(el).attr('href'))),
      // legacy
      ...$('script').filter((_, el) => Boolean($(el).data('src') && $(el).attr('id') === 'vite-legacy-entry')),
      // log-time
      ...$('script').filter((_, el) => typeof $(el).data('log-time') !== 'undefined'),
    ]
  }

  const dynamicTags = generateDynamicTags($)
  const dynamicTagsOfPre = generateDynamicTags(pre$)

  dynamicTags.forEach((el) => {
    $(el).remove()
  })

  const createInsertScript = (dynamicTags: Element[]) => {
    return dynamicTags.map((e) => ({
      ...e.attribs,
      'data-tag': e.tagName,
    }))
  }

  const code = /* js */ `
    !(function() {
      function insertTags(attrs) {
        attrs.forEach((attr) => {
          const element = document.createElement(attr['data-tag'])
          for (const key in attr) {
            element.setAttribute(key, attr[key])
          }
          document.head.appendChild(element)
        })
      }
      const isPrerelease = window.Cookies.get('prerelease') === 'true'
      const attrs = isPrerelease ? ${JSON.stringify(createInsertScript(dynamicTagsOfPre))} : ${JSON.stringify(createInsertScript(dynamicTags))}
      insertTags(attrs)
    })()
  `

  $('body').after(`<script>${code}</script>`)

  return $.html().replace(/^\s*$(?:\r\n?|\n)/gm, '')
}
