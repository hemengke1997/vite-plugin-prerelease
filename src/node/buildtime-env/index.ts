import fsp from 'node:fs/promises'
import path from 'node:path'
import { build, normalizePath, type PluginOption, type ResolvedConfig } from 'vite'
import { type Options } from '..'
import { transformHtml } from './utils'

export function buildtimeEnv(options: Pick<Required<Options>, 'prereleaseEnv' | '__debug'>): PluginOption[] {
  const { prereleaseEnv, __debug } = options || {}

  // Define environment variables when building pre-release
  const definePlugin: PluginOption = {
    name: 'vite:plugin-prerelease-define',
    apply(_, env) {
      return (env.mode === prereleaseEnv && env.command === 'build') || __debug
    },
    config() {
      return {
        define: {
          'import.meta.env.PRERELEASE': 'true',
        },
      }
    },
  }

  let config: ResolvedConfig
  let originalHtml: string
  let prereleaseHtml: string
  // Build pre-release after building test
  const buildPlugin: PluginOption = {
    name: 'vite:plugin-prerelease-build',
    enforce: 'post',
    apply(_, env) {
      if (env.mode === prereleaseEnv) return false
      return env.command === 'build'
    },
    configResolved(resolvedConfig) {
      config = resolvedConfig
    },
    transformIndexHtml: {
      order: 'post',
      handler(html) {
        originalHtml = html
      },
    },
    async closeBundle() {
      config.logger.info('Building prerelease...')

      await build({
        configFile: config.configFile,
        mode: prereleaseEnv,
        base: normalizePath(`${config.base}/prerelease`),
        build: {
          outDir: normalizePath(`${config.build.outDir}/prerelease`),
        },
        plugins: [
          {
            name: 'vite:plugin-prerelease-html',
            enforce: 'post',
            transformIndexHtml: {
              order: 'post',
              handler(html) {
                prereleaseHtml = html
              },
            },
          },
        ],
      })

      const transformedHtml = transformHtml(originalHtml, prereleaseHtml)
      if (transformedHtml) {
        await fsp.writeFile(path.join(config.build.outDir, 'index.html'), transformedHtml)
      }
      config.logger.info('Prerelease build complete')
    },
  }

  return [definePlugin, buildPlugin]
}
