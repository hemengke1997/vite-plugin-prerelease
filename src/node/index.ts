import fsp from 'node:fs/promises'
import { createRequire } from 'node:module'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import serialize from 'serialize-javascript'
import {
  build,
  type ConfigEnv,
  type HtmlTagDescriptor,
  normalizePath,
  type PluginOption,
  type ResolvedConfig,
  transformWithEsbuild,
} from 'vite'
import { type PrereleaseWidgetOptions } from '../client/core/types'
import { transformHtml } from './utils'

type Options = {
  /**
   * 预发布环境名称
   * @default 'production'
   */
  prereleaseEnv?: string

  /**
   * 预发布小组件配置
   */
  prereleaseWidget?: PrereleaseWidgetOptions

  /**
   * debug
   * @internal
   */
  __debug?: boolean
}

declare global {
  interface ImportMetaEnv {
    PRERELEASE: string
  }
  interface Window {
    Cookies: typeof import('js-cookie')
  }
}

export function prerelease(options?: Options): any {
  const { prereleaseEnv = 'production', prereleaseWidget = {}, __debug = false } = options || {}

  if (process.env.NODE_ENV === prereleaseEnv || process.env.NODE_ENV === 'production') {
    return
  }

  // Insert the pre-release widget when building test and pre-release
  let env: ConfigEnv
  const configPlugin: PluginOption = {
    name: 'vite:plugin-prerelease-config',
    enforce: 'pre',
    apply(_, env) {
      return env.command === 'build' || __debug
    },
    config(_, configEnv) {
      env = configEnv
    },
    transformIndexHtml: {
      // pre for base url
      order: 'pre',
      async handler(html) {
        const require = createRequire(import.meta.url)
        const __dirname = path.dirname(fileURLToPath(import.meta.url))

        const prereleaseWidgetPath = normalizePath(`/@fs/${path.join(__dirname, './client/index.js')}`)

        const prereleaseWidgetScript = /*js*/ `
          import { PrereleaseWidget } from '${prereleaseWidgetPath}'
          if(typeof window !== 'undefined' && typeof document !== 'undefined') {
            const timer = setTimeout(() => {
              new PrereleaseWidget(${serialize(prereleaseWidget)})
            })
          }
        `

        const tags: HtmlTagDescriptor[] = [
          {
            tag: 'script',
            injectTo: 'body',
            attrs: {
              type: 'module',
            },
            children: prereleaseWidgetScript,
          },
        ]

        if (env.mode !== prereleaseEnv) {
          const jsCookie = await transformWithEsbuild(
            await fsp.readFile(path.join(path.dirname(require.resolve('js-cookie')), 'js.cookie.js'), 'utf-8'),
            '',
          )

          tags.push({
            tag: 'script',
            injectTo: 'head-prepend',
            children: jsCookie.code,
          })
        }

        return {
          html,
          tags,
        }
      },
    },
  }

  // Define environment variables when building pre-release
  const definePlugin: PluginOption = {
    name: 'vite:plugin-prerelease-define',
    apply(_, env) {
      return (env.mode === prereleaseEnv && env.command === 'build') || __debug
    },
    config() {
      return {
        define: {
          'import.meta.env.PRERELEASE': 'true',
        },
      }
    },
  }

  let config: ResolvedConfig
  let originalHtml: string
  let prereleaseHtml: string
  // Build pre-release after building test
  const buildPlugin: PluginOption = {
    name: 'vite:plugin-prerelease-build',
    enforce: 'post',
    apply(_, env) {
      if (env.mode === prereleaseEnv) return false
      return env.command === 'build'
    },
    configResolved(resolvedConfig) {
      config = resolvedConfig
    },
    transformIndexHtml: {
      order: 'post',
      handler(html) {
        originalHtml = html
      },
    },
    async closeBundle() {
      config.logger.info('Building prerelease...')

      await build({
        configFile: config.configFile,
        mode: prereleaseEnv,
        base: normalizePath(`${config.base}/prerelease`),
        build: {
          outDir: normalizePath(`${config.build.outDir}/prerelease`),
        },
        plugins: [
          {
            name: 'vite:plugin-prerelease-html',
            transformIndexHtml: {
              order: 'post',
              handler(html) {
                prereleaseHtml = html
              },
            },
          },
        ],
      })

      const transformedHtml = transformHtml(originalHtml, prereleaseHtml)
      await fsp.writeFile(path.join(config.build.outDir, 'index.html'), transformedHtml)
      config.logger.info('Prerelease build complete')
    },
  }

  return [configPlugin, definePlugin, buildPlugin]
}
