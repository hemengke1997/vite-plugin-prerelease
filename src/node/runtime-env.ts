import MagicString from 'magic-string'
import { type PluginOption, type ResolvedConfig } from 'vite'
import { type Options } from '.'
import { resolveEnvFromConfig, runtimeEnvCode } from './utils'

export function runtimeEnv(options: Pick<Required<Options>, 'prereleaseEnv'>): PluginOption[] {
  const { prereleaseEnv } = options

  let env: ReturnType<typeof resolveEnvFromConfig>

  let config: ResolvedConfig

  const runtimeEnvPlugin: PluginOption = {
    name: 'vite:plugin-runtime-env',
    async configResolved(_config) {
      config = _config

      env = resolveEnvFromConfig(config, prereleaseEnv)
    },
    transform(code, _, options) {
      const { ssr } = options || {}

      const magicString = new MagicString(code)

      const importMetaPattern = /import\.meta\.env\.([A-Z0-9_]+)/g
      let match: RegExpExecArray | null

      while ((match = importMetaPattern.exec(code))) {
        const start = match.index
        const end = start + match[0].length
        const name = match[1]

        magicString.overwrite(start, end, `${ssr ? 'global' : 'window'}.__env__?.${name}`)
      }

      if (!magicString.hasChanged()) {
        return {
          code,
          map: null,
        }
      }

      if (!config.build.sourcemap) {
        return {
          code: magicString.toString(),
          map: null,
        }
      }

      return {
        code: magicString.toString(),
        map: magicString.generateMap({ hires: true }),
      }
    },
    transformIndexHtml() {
      return [
        {
          tag: 'script',
          attrs: { type: 'application/javascript' },
          injectTo: 'body',
          children: runtimeEnvCode(env),
        },
      ]
    },
  }

  return [runtimeEnvPlugin]
}
