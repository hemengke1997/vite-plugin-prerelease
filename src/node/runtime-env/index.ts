import MagicString from 'magic-string'
import { type PluginOption, type ResolvedConfig } from 'vite'
import { type Options } from '..'
import { resolveEnvFromConfig, runtimeEnvCode } from './utils'

export function runtimeEnv(options: Pick<Required<Options>, 'prereleaseEnv' | 'excludeEnvs'>): PluginOption[] {
  const { prereleaseEnv, excludeEnvs } = options

  let env: ReturnType<typeof resolveEnvFromConfig>

  let config: ResolvedConfig

  const runtimeEnvPlugin: PluginOption = {
    name: 'vite:plugin-runtime-env',
    async configResolved(_config) {
      config = _config

      env = resolveEnvFromConfig(config, prereleaseEnv)
    },
    transform(code, _, options) {
      const { ssr } = options || {}

      const magicString = new MagicString(code)

      const importMetaPattern = /import\.meta\.env(?:\.([A-Z0-9_]+))?/g
      let match: RegExpExecArray | null

      while ((match = importMetaPattern.exec(code))) {
        const start = match.index
        const end = start + match[0].length
        const name = match[1]

        if (name && excludeEnvs.includes(name)) {
          continue
        }

        magicString.overwrite(
          start,
          end,
          /*js*/ `(() => { 
              let isPreRelease = false
              if(typeof window !== 'undefined' && window.Cookies?.get('prerelease') === 'true') {
                isPreRelease = true
              }
              if(typeof global !== 'undefined' && global.__isPrerelease__) {
                isPreRelease = true
              }
              return ${ssr ? 'global' : 'window'}.__env__?.[isPreRelease ? 'prerelease' : 'current']${name ? `.${name}` : ''}
            })()`,
        )
      }

      if (!magicString.hasChanged()) {
        return {
          code,
          map: null,
        }
      }

      if (!config.build.sourcemap) {
        return {
          code: magicString.toString(),
          map: null,
        }
      }

      return {
        code: magicString.toString(),
        map: magicString.generateMap({ hires: true }),
      }
    },
    transformIndexHtml() {
      return [
        {
          tag: 'script',
          attrs: { type: 'application/javascript' },
          injectTo: 'head',
          children: runtimeEnvCode(env),
        },
      ]
    },
  }

  return [runtimeEnvPlugin]
}
