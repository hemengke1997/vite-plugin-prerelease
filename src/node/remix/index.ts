import MagicString from 'magic-string'
import serialize from 'serialize-javascript'
import ts from 'typescript'
import { type Plugin, type ResolvedConfig } from 'vite'
import { type Options } from '../index'
import { prerelease as main } from '../index'
import { resolveEnvFromConfig, runtimeEnvCode } from '../runtime-env/utils'
import { resolveJsCookie } from '../utils'
import { resolvedVirtualModuleId, runtimeId, serverId, vmods } from './virtual'

export function prerelease(options?: Pick<Options, 'prereleaseWidget' | 'prereleaseEnv'>): any {
  const { prereleaseWidget = {}, prereleaseEnv = 'production' } = options || {}
  if (process.env.NODE_ENV === prereleaseEnv || process.env.NODE_ENV === 'production') {
    return
  }

  let config: ResolvedConfig
  let env: ReturnType<typeof resolveEnvFromConfig>

  return [
    {
      name: 'vite:plugin-prerelease-remix',
      enforce: 'pre',
      configResolved: {
        order: 'pre',
        handler(_config) {
          config = _config
          env = resolveEnvFromConfig(config, prereleaseEnv)
        },
      },
      transform: {
        order: 'pre',
        async handler(code, id) {
          if (/.+\/root\.[t|j]sx?$/.test(id)) {
            const app = 'AppExportIdentifier'

            const s = new MagicString(code)
            const sourceFile = ts.createSourceFile('source.tsx', code, ts.ScriptTarget.Latest, true, ts.ScriptKind.TSX)

            function visit(node: ts.Node) {
              if (ts.isFunctionDeclaration(node)) {
                const defaultNode = node.modifiers?.find((m) => m.kind === ts.SyntaxKind.DefaultKeyword)
                if (defaultNode) {
                  s.remove(defaultNode.pos, defaultNode.end)
                  if (node.name) {
                    s.overwrite(node.name.pos + 1, node.name.end, app)
                  } else {
                    s.prependLeft(node.pos, `\nconst ${app} =`)
                  }
                }
              } else if (ts.isExportAssignment(node)) {
                s.overwrite(node.pos, node.expression.pos, `\nconst ${app} =`)
              }
              ts.forEachChild(node, visit)
            }

            visit(sourceFile)

            const imports = [/*js*/ `import { withPrereleaseWidget } from "vite-plugin-prerelease/remix/client";`]
            const defaultExport = /*js*/ `export default withPrereleaseWidget(${app}, ${serialize(prereleaseWidget)})();`

            code = `${imports.join('\n')}\n${s.toString()}\n${defaultExport}`

            return {
              code,
              map: null,
            }
          }
        },
      },
      resolveId(id) {
        if (vmods.includes(id)) {
          return resolvedVirtualModuleId(id)
        }
        return null
      },
      async load(id) {
        switch (id) {
          case resolvedVirtualModuleId(runtimeId): {
            const jsCookie = await resolveJsCookie()
            const envCode = runtimeEnvCode(env)
            return /*js*/ `
              if(typeof window !== 'undefined') {
                ${jsCookie}
                ${envCode}
              }
            `
          }
          case resolvedVirtualModuleId(serverId): {
            return /*js*/ `
              import cookie from 'cookie'
              export default function server(request) {
                const cookies = cookie.parse(request.headers.get('Cookie') ?? '')
                global.isPreRelease = cookies.prerelease === 'true'
              }
            `
          }
          default:
            break
        }
      },
    },
    main({
      mode: 'runtime',
      prereleaseEnv,
      prereleaseWidget,
    }),
  ] as Plugin[]
}
