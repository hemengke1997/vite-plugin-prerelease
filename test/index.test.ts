import fs from 'node:fs'
import path from 'node:path'
import { describe, expect, test } from 'vitest'
import { transformHtml } from '../src/node/utils'

function transform() {
  const originHtml = fs.readFileSync(path.join(__dirname, './fixtures/test.html'), 'utf-8')
  const prereleaseHtml = fs.readFileSync(path.join(__dirname, './fixtures/prerelease.html'), 'utf-8')

  const result = transformHtml(originHtml, prereleaseHtml)
  return result
}

describe('transform html', () => {
  test('should match snapshot', () => {
    const result = transform()

    expect(result).toMatchSnapshot()
  })
  test('should transform match result', async () => {
    const result = transform()

    await expect(result).toMatchFileSnapshot('./fixtures/index.html')
  })
})
