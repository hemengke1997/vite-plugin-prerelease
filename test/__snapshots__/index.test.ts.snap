// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform html > should match snapshot 1`] = `
"<!DOCTYPE html><html lang="en" data-theme="dark"><head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>vite-plugin-prerelease</title>
  <script type="module">import.meta.url; import("_").catch(() => 1); (async function* () { })().next(); if (location.protocol != "file:") { window.__vite_is_modern_browser = true }</script>
  <script type="module">!function () { if (window.__vite_is_modern_browser) return; console.warn("vite: loading legacy chunks, syntax error above and the same error below should be ignored"); var e = document.getElementById("vite-legacy-polyfill"), n = document.createElement("script"); n.src = e.src, n.onload = function () { System.import(document.getElementById('vite-legacy-entry').getAttribute('data-src')) }, document.body.appendChild(n) }();</script>
</head>
<body>
  <div id="root"></div>
  <script nomodule="">!function () { var e = document, t = e.createElement("script"); if (!("noModule" in t) && "onbeforeload" in t) { var n = !1; e.addEventListener("beforeload", (function (e) { if (e.target === t) n = !0; else if (!e.target.hasAttribute("nomodule") || !n) return; e.preventDefault() }), !0), t.type = "module", t.src = ".", e.head.appendChild(t), t.remove() } }();</script>
</body><script>
    !(function() {
      function insertTags(attrs) {
        attrs.forEach((attr) => {
          const element = document.createElement(attr['data-tag'])
          for (const key in attr) {
            element.setAttribute(key, attr[key])
          }
          document.head.appendChild(element)
        })
      }
      const isPrerelease = window.Cookies.get('prerelease') === 'true'
      const attrs = isPrerelease ? [{"type":"module","crossorigin":"","src":"/vite-plugin-prerelease/prerelease/assets/js/polyfills-DlnDHItd.js","data-tag":"script"},{"type":"module","crossorigin":"","src":"/vite-plugin-prerelease/prerelease/assets/js/index-D3bJ_PZJ.js","data-tag":"script"},{"nomodule":"","crossorigin":"","id":"vite-legacy-polyfill","src":"/vite-plugin-prerelease/prerelease/assets/js/polyfills-legacy-B7oU9mq8.js","data-tag":"script"},{"rel":"icon","type":"image/svg+xml","href":"/vite-plugin-prerelease/prerelease/vite.svg","data-tag":"link"},{"rel":"stylesheet","crossorigin":"","href":"/vite-plugin-prerelease/prerelease/assets/css/index-Cx8BHIrc.css","data-tag":"link"},{"nomodule":"","crossorigin":"","id":"vite-legacy-entry","data-src":"/vite-plugin-prerelease/prerelease/assets/js/index-legacy-BzAYR8n5.js","data-tag":"script"},{"data-log-time":"","data-tag":"script"}] : [{"type":"module","crossorigin":"","src":"/vite-plugin-prerelease/assets/js/polyfills-Cxu7lcET.js","data-tag":"script"},{"type":"module","crossorigin":"","src":"/vite-plugin-prerelease/assets/js/index-KsqBbYBB.js","data-tag":"script"},{"nomodule":"","crossorigin":"","id":"vite-legacy-polyfill","src":"/vite-plugin-prerelease/assets/js/polyfills-legacy-XqQan7A1.js","data-tag":"script"},{"rel":"icon","type":"image/svg+xml","href":"/vite-plugin-prerelease/vite.svg","data-tag":"link"},{"rel":"stylesheet","crossorigin":"","href":"/vite-plugin-prerelease/assets/css/index-ZB8290Ql.css","data-tag":"link"},{"nomodule":"","crossorigin":"","id":"vite-legacy-entry","data-src":"/vite-plugin-prerelease/assets/js/index-legacy-Loeqjk_U.js","data-tag":"script"},{"data-log-time":"","data-tag":"script"}]
      insertTags(attrs)
    })()
  </script></html>"
`;
